TabnineSelfHosted/src/main/java/com/tabnineSelfHosted/statusBar/TabnineSelfHostedStatusBarWidget.kt

```diff
@@ -14,38 +14,56 @@ import com.tabnineCommon.binary.requests.config.CloudConnectionHealthStatus
 import com.tabnineCommon.binary.requests.config.StateResponse
 import com.tabnineCommon.general.StaticConfig
 import com.tabnineCommon.lifecycle.BinaryStateChangeNotifier
+import com.tabnineCommon.userSettings.AppSettingsState
+import com.tabnineSelfHosted.binary.lifecycle.UserInfoChangeNotifier
+import com.tabnineSelfHosted.binary.requests.userInfo.UserInfoResponse
 import com.tabnineSelfHosted.statusBar.SelfHostedStatusBarActions.buildStatusBarActionsGroup
 import java.awt.event.MouseEvent
 import javax.swing.Icon
 
-class TabnineSelfHostedStatusBarWidget(project: Project) : EditorBasedWidget(project), StatusBarWidget, MultipleTextValuesPresentation {
+class TabnineSelfHostedStatusBarWidget(project: Project) :
+    EditorBasedWidget(project),
+    StatusBarWidget,
+    MultipleTextValuesPresentation {
     private var cloudConnectionHealthStatus = CloudConnectionHealthStatus.Ok
-    private var username: String? = null
+    private var userInfo: UserInfoResponse? = null
 
     init {
-        // register for state changes (we will get notified whenever the state changes)
         ApplicationManager.getApplication()
             .messageBus
             .connect(this)
             .subscribe(
                 BinaryStateChangeNotifier.STATE_CHANGED_TOPIC,
-                BinaryStateChangeNotifier { (_, _, _, cloudConnectionHealthStatus1, _, userName): StateResponse ->
-                    cloudConnectionHealthStatus = cloudConnectionHealthStatus1
-                    username = userName
+                BinaryStateChangeNotifier { (_, _, _, cloudConnectionHealthStatus, _, _): StateResponse ->
+                    this.cloudConnectionHealthStatus = cloudConnectionHealthStatus
+                    update()
+                }
+            )
+
+        ApplicationManager.getApplication()
+            .messageBus
+            .connect(this)
+            .subscribe(
+                UserInfoChangeNotifier.USER_INFO_CHANGED_TOPIC,
+                UserInfoChangeNotifier { userInfoResponse ->
+                    userInfo = userInfoResponse
```

Comment:
I don't like that you're saving all the states coming from the notifiers (the `cloudConnectionHealthStatus` and the `userInfo` fields) - can't you use the `UserInfoService` for this? It already saves the last user info response.
The same goes for the `cloudConnectionHealthStatus` - I can't remember if it has a matching service similar to the `UserInfoService`, and if it not, it might be worth creating one.

going by your implementation is dictating a bad design, where every consumer of these states saves a copy of the same response.